### DISCUSSION TITLE
DependencyInjection 还有其他妙用吗？

### DISCUSSION THREAD

**Discussion by ZhuYajunFly** [BYJRK/DotNet-Discussions#337] [C# 技术交流]
Microsoft.Extensions.DependencyInjection后面简称DependencyInjection.
DependencyInjection一般除了获取View或者View Model或者某些类的引用，注册为AddSingleton或者AddTransient等这样的方式，还有其他妙用吗？
另外还有一个期盼的功能，比如一个普通类，我想给容器注册两个不同引用的单例实例，就是这两个单例实例我在app任何地方都能拿到，取值赋值，能做到不？这两个单例实例，或许叫全局实例更妥当一些。

**Comment by BYJRK** [comment]
你其实搞反了😂
DI 本来的作用是在诸如 ASP.NET Core 等后端框架中用于注册各种服务、中间件等。只是大家发现它还可以用在 WPF 中，从而把 View 及 ViewModel 也注册进去，从而方便 IoC 容器去处理它们的依赖以及生命周期，所以才有了这种“玩法”。
但是大家还发现，直接把 DI 用在 WPF 中，仍然存在一定的局限性，就比如想要实现导航功能的时候，因为 WPF 不能像前端那样直接以“网址”的形式导航到页面，以及传参，还有借助 Scoped 生命周期来维持一些对象的状态，所以往往还需要在 WPF 中实现诸如 ViewLocator、NavigationService 等类。

**Reply by ZhuYajunFly** [reply]
> 你其实搞反了😂 DI 本来的作用是在诸如 ASP.NET Core 等后端框架中用于注册各种服务、中间件等。只是大家发现它还可以用在 WPF 中，从而把 View 及 ViewModel 也注册进去，从而方便 IoC 容器去处理它们的依赖以及生命周期，所以才有了这种“玩法”。 但是大家还发现，直接把 DI 用在 WPF 中，仍然存在一定的局限性，就比如想要实现导航功能的时候，因为 WPF 不能像前端那样直接以“网址”的形式导航到页面，以及传参，还有借助 Scoped 生命周期来维持一些对象的状态，所以往往还需要在 WPF 中实现诸如 ViewLocator、NavigationService 等类。

是嘛，大佬。不过有个好处是，View 及 ViewModel 也注册成单例，有时候迫不得已去在ViewModel中获取View中的控件(比如著名的图表库ScottPlot不支持绑定，需要更新曲线时)，确实方便。
我想把一个类注册多个不同的单例，这个可以不，比如我一份数据，整个生命周期存在两个份，可能不同的地方都会去访问这两份数据。

**Comment by BYJRK** [comment]
你的另一个需求太奇怪了。你既然声明了两个“单例”，那么你总会有一种方式来区分你在某种情况下到底想要哪一个吧。那么你可以考虑使用不同的接口来注册服务，或者使用 DI 最新的 `KeyedService` 功能来使用字符串关键字来注册并获取服务。

**Reply by ZhuYajunFly** [reply]
> 你的另一个需求太奇怪了。你既然声明了两个“单例”，那么你总会有一种方式来区分你在某种情况下到底想要哪一个吧。那么你可以考虑使用不同的接口来注册服务，或者使用 DI 最新的 `KeyedService` 功能来使用字符串关键字来注册并获取服务。

单例只是形容这两个独立的实例，能够在不同的VM、ViewModel去访问，我想的就类似一Message通过一个String类型的token区分。刚刚测试了一下，整两个容器，各自去注册一份，public IServiceProvider Services2 { get; }，这样应该也可以可以。最新的KeyedService有没有示例代码呀。

**Reply by ZhuYajunFly** [reply]
> 你的另一个需求太奇怪了。你既然声明了两个“单例”，那么你总会有一种方式来区分你在某种情况下到底想要哪一个吧。那么你可以考虑使用不同的接口来注册服务，或者使用 DI 最新的 `KeyedService` 功能来使用字符串关键字来注册并获取服务。

感谢大佬，AddKeyedSingleton确实是可以注册多个单例的。
![image](https://github.com/user-attachments/assets/a34dc948-0af8-4d0f-8b2d-8a5ccce10916)
![image](https://github.com/user-attachments/assets/9e514dd5-1966-4ba9-b157-3594b45762d8)
![image](https://github.com/user-attachments/assets/5efe06a7-ecff-4194-8484-c9af93c4ebf0)
